websocket.go

```diff
@@ -0,0 +1,277 @@
+					unsubscribeMsg := map[string]any{
+						"op":      "unsubscribe",
+						"channel": "swap.update",
+						"args":    []string{swapID},
+					}
+					messageBytes, err := json.Marshal(unsubscribeMsg)
+					if err != nil {
+						log.Printf("Failed to marshal unsubscribe message: %v", err)
+						continue
+					}
+					err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+					if err != nil {
+						log.Printf("Failed to forward unsubscribe message to upstream: %v", err)
+					}
+				}
+			}
+		}
+
+		// Remove the client from the clients map
+		delete(p.clients, conn)
+	}()
+
+	for {
+		// Read message from client with timeout
+		_, message, err := conn.Read(ctx)
+		if err != nil {
+			if websocket.CloseStatus(err) == websocket.StatusNormalClosure {
+				log.Printf("Client closed the connection: %v", err)
+			} else {
+				log.Printf("WebSocket read error: %v", err)
+			}
+			break
+		}
+
+		// Parse the message
+		var msg map[string]any
+		if err := json.Unmarshal(message, &msg); err != nil {
+			log.Printf("Failed to parse client message: %v", err)
+			continue
+		}
+
+		// Handle subscription/unsubscription
+		switch msg["op"] {
+		case "subscribe":
+			p.handleSubscribe(conn, msg)
+		case "unsubscribe":
+			p.handleUnsubscribe(conn, msg)
+		default:
+			log.Printf("Unknown operation: %s", msg["op"])
+		}
+
+		// Reset the timeout for the next read
+		cancel() // Cancel the previous context
+		ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel() // Ensure the new context is canceled on all paths
+	}
+}
+
+func (p *WebSocketProxy) handleSubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid subscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for _, swapID := range swapIDs {
+		id, ok := swapID.(string)
+		if !ok {
+			log.Printf("Invalid swap ID: %v", swapID)
+			continue
+		}
+
+		// Add client to swap ID's subscribers
+		if p.subscribers[id] == nil {
+			p.subscribers[id] = make(map[*websocket.Conn]bool)
+		}
+		p.subscribers[id][conn] = true
+
+		// Add swap ID to client's subscriptions
+		p.clients[conn][id] = true
+	}
+
+	// Forward subscribe message to upstream
+	if p.upstream != nil {
+		// Marshal the message into JSON ([]byte)
+		messageBytes, err := json.Marshal(msg)
+		if err != nil {
+			log.Printf("Failed to marshal subscribe message: %v", err)
+			return
+		}
+
+		// Write the message to the upstream WebSocket
+		err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+		if err != nil {
+			log.Printf("Failed to forward subscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) handleUnsubscribe(conn *websocket.Conn, msg map[string]any) {
+	swapIDs, ok := msg["args"].([]any)
+	if !ok {
+		log.Printf("Invalid unsubscribe message: %v", msg)
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
```

Comment:
This unlock can go below the for loop, before the `p.upstream != nil`, so the lock is not taken during the upstream write.